eos-paygd/main.c

```diff
@@ -35,8 +35,92 @@
 #include <sys/un.h>
 #include <libeos-payg/efi.h>
 
+#define LOGFILE_DIRNAME "/var/log/eos-payg"
+#define LOGFILE_BASENAME "eos-paygd"
+#define LOGFILE_EXT "log"
+
 #define TIMEOUT_POWEROFF_ON_ERROR_MINUTES 20
 
+static gboolean
+open_log_file (GIOChannel **log_io)
+{
+  const gchar *log_file_name = NULL;
+  g_autoptr(GFile) log_file = NULL;
+  g_autoptr(GDateTime) now = NULL;
+  g_autofree gchar *tstamp = NULL;
+
+  /* Build log file name with a date stamp so we get one file per day,
+   * ex., eos-paygd-20220418.log
+   */
+  now = g_date_time_new_now_local ();
+  tstamp = g_date_time_format (now, "%Y%m%d");
+  log_file_name = g_strconcat (LOGFILE_DIRNAME, "/", LOGFILE_BASENAME, "-",
+                               tstamp, ".", LOGFILE_EXT, NULL);
+  log_file = g_file_new_for_path (log_file_name);
+
+  *log_io = g_io_channel_new_file (log_file_name, "a", NULL);
+  if (!log_io)
```

Comment:
You mean `if (!*log_io)` here. This function as written will either dereference a null pointer, or return TRUE unconditionally.

You could choose to instead return the `GIOChannel *` rather than `gboolean` and have no out parameter, which would make this bug harder to write.

Reply:
Oh, good catch, thanks! Yeah, I usually don't like out parameters because it is easy to overlook the dereferencing. There was something on the previous versions that made me think it was better that way, but with the way the code is right now I agree ditching the out parameter entirely.